java/ql/src/Security/CWE/CWE-90/LdapInjection.qhelp

```diff
@@ -0,0 +1,62 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>If an LDAP query is built using string concatenation, and the
+components of the concatenation include user input, a user 
+is likely to be able to run malicious LDAP queries.</p>
+</overview>
+
+<recommendation>
+<p>If user input must be included in an LDAP query, it should be escaped to
+avoid a malicious user providing special characters that change the meaning
+of the query. If possible build the LDAP query (or search filter / DN) using your
+framework helper methods to avoid string concatenation, or escape user input
+using the right LDAP encoding method, for example encodeForLDAP from OWASP ESAPI,
+LdapEncoder from Spring LDAP or Filter.encodeValue from UnboundID library.</p>
```

Comment:
The second sentence here is rather long. I'd suggest simplifying as follows:

> If possible use framework helper methods instead of string concatenation to build the search filter, or distinguished name (DN), for the LDAP query. Alternatively, escape user input using an appropriate LDAP encoding method, for example: <code>encodeForLDAP</code> from OWASP ESAPI, <code>LdapEncoder</code> from Spring LDAP, or <code>Filter.encodeValue</code> from UnboundID library.

**Note:** I originally read line 14 as meaning that you should build the LDAP query, the search filter, or the DN using a framework helper method. However, when I read the rest of the help topic I realized I'd misunderstood. I *think* that we're talking about an LDAP query that might contain a search filter and/or a DN. If my understanding (and therefore my suggestion) is still wrong, please clarify the intention of line 14.

Reply:
I agree that the sentence is unclear. Let me clarify what I meant.
The best recommendation is to build LDAP query using the framework's helper methods (for example from Spring's `LdapQueryBuilder` and `LdapNameBuilder`), not string concatenation:
```java
LdapQuery query = query()
  .base(LdapNameBuilder.newInstance("dc=261consulting,dc=com").add("ou", "people").build())
  .searchScope(SearchScope.ONELEVEL)
  .timeLimit(200)
  .countLimit(221)
  .where("objectclass").is("person").and("cn").is("John Doe");
```
In such case, individual filter/DN parts don't need to be escaped.
If using helper methods is not an option, the LDAP search filter and DN should be escaped with `encodeForLDAP` / `encodeForDN` from OWASP ESAPI, `LdapEncoder.filterEncode`/`LdapEncoder.nameEncode` from Spring LDAP, or `Filter.encodeValue` from UnboundID library.
Maybe the second sentence could be changed to:
> If possible use framework helper methods, for example from Spring's `LdapQueryBuilder` and `LdapNameBuilder`, instead of string concatenation to build the LDAP query. Alternatively, escape user input using an appropriate LDAP encoding method, for example: `encodeForLDAP` or `encodeForDN` from OWASP ESAPI, `LdapEncoder.filterEncode` or `LdapEncoder.nameEncode` from Spring LDAP, or `Filter.encodeValue` from UnboundID library.

Reply:
Thanks for clarifying this. Your new version is much clearer but I think the best solution is to also revise the first sentence so it uses your original sentence order:

> If possible build the LDAP query using framework helper methods, for example from Spring's LdapQueryBuilder and LdapNameBuilder, instead of string concatenation. Alternatively, ...

Thanks for your patience with this.

Reply:
I applied your suggestion. Please double check if everything is OK.
Thank you for your patience and for correcting me :)

Reply:
Thanks. I think this is clear now. Unfortunately the `qhelp` preview test failed on a minor typo, but as far as I can tell, the text all looks fine now. 😄